### PR TITLE
add smali language support

### PR DESCRIPTION
**Pull request by evanrichter**

### CODE REVIEW

File: runtime/queries/smali/highlights.scm [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+   . (identifier) @_first @type.builtin
+   .  (identifier) @_second @type.builtin
+   (identifier) @type.builtin)
+  (#eq? @_first "com")
+  (#any-of? @_second "android" "google"))
+
+; Methods
+
+(method_definition
+  (method_signature (method_identifier) @method))
```
Comment (archseer) [inline runtime/queries/smali/highlights.scm:23]: Quite a few scopes in this file don't match our expected scopes so they won't highlight correctly: https://docs.helix-editor.com/themes.html#syntax-highlighting

Comment (evanrichter) [reply runtime/queries/smali/highlights.scm:23]: Ok I'll work on it!

I just copied these scm files from upstream and it looked good enough to me in the smali files I am editing

Comment (evanrichter) [reply runtime/queries/smali/highlights.scm]: Should be much better now! I left my changes as a separate commit for review, but I can easily rebase this PR into one commit if you prefer

File: runtime/queries/smali/highlights.scm [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+  (#any-of? @constructor "<init>" "<clinit>"))
+
+"constructor" @constructor
+
+; Fields
+
+[
+  (field_identifier)
+  (annotation_key)
+] @field
```
Comment (the-mikedavis) [inline runtime/queries/smali/highlights.scm:74]:
```suggestion
] @variable.other.member
```

File: runtime/queries/smali/highlights.scm [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+  ".catchall"
+] @keyword.control.exception
+
+; Literals
+
+(string) @string
+(source_directive (string "\"" _ @string.special.url "\""))
+(escape_sequence) @constant.character.escape
+
+(character) @character
```
Comment (the-mikedavis) [inline runtime/queries/smali/highlights.scm:170]:
```suggestion
(character) @constant.character
```

File: runtime/queries/smali/highlights.scm [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+   . (identifier) @_first @type.builtin
+   (identifier) @type.builtin)
+  (#any-of? @_first "android" "dalvik" "java" "kotlinx"))
+
+((class_identifier
+   . (identifier) @_first @type.builtin
+   . (identifier) @_second @type.builtin
+   (identifier) @type.builtin)
+  (#eq? @_first "com")
+  (#any-of? @_second "android" "google"))
```
Comment (the-mikedavis) [inline runtime/queries/smali/highlights.scm:18]: We don't currently support `#any-of?`. These will need to be rewritten to use `#match?` instead, for example `(#match? @_second "^(android|google)$")`

File: runtime/queries/smali/injections.scm [hunk 5]
```diff
@@ -0,0 +1 @@
+(comment) @comment
```
Comment (the-mikedavis) [inline runtime/queries/smali/injections.scm:1]:
```suggestion
((comment) @injection.content
 (#set! injection.language "comment"))
```

we follow tree-sitter for injections https://tree-sitter.github.io/tree-sitter/syntax-highlighting#language-injection, nvim has its own injection format

File: runtime/queries/smali/locals.scm
Comment (the-mikedavis) [inline runtime/queries/smali/locals.scm:1]: We also follow tree-sitter for locals: https://tree-sitter.github.io/tree-sitter/syntax-highlighting#local-variables

So these need to be rewritten to only use the captures `@local.scope`, `@local.reference` and `@local.definition`

File: runtime/queries/smali/indents.scm
Comment (the-mikedavis) [inline runtime/queries/smali/indents.scm:1]: These look like nvim indent queries. See https://docs.helix-editor.com/master/guides/indent.html for writing Helix ones

File: languages.toml [hunk 6]
```diff
@@ -843,6 +843,19 @@ indent = { tab-width = 2, unit = "  " }
 name = "java"
 source = { git = "https://github.com/tree-sitter/tree-sitter-java", rev = "09d650def6cdf7f479f4b78f595e9ef5b58ce31e" }
 
+[[language]]
+name = "smali"
+scope = "source.smali"
+injection-regex = "smali"
+file-types = ["smali"]
+comment-token = "#"
+language-servers = []
```
Comment (the-mikedavis) [inline languages.toml:852]: This is the default so this key can be removed

File: runtime/queries/smali/indents.scm
Comment (evanrichter) [reply runtime/queries/smali/indents.scm:1]: ok I did my best! smali does not use indentation too much (it's basically just a text representation of dalvik bytecode with gotos for control flow). But I checked the behavior with this example with the most indentation I could find and it seems correct:

```smali
.method public static updateDigest(Ljava/security/MessageDigest;Ljava/io/File;)V
    .locals 2
    .annotation system Ldalvik/annotation/Throws;
        value = {
            Ljava/io/IOException;
        }
    .end annotation
```

File: runtime/queries/smali/injections.scm (hunk 5 above)
Comment (evanrichter) [reply runtime/queries/smali/injections.scm]: I made the change as suggested, and this reminded me to check `goto_next_comment` which I found doesn't work.

tree-sitter-subtree does report `(comment)` for comment lines, so I don't know what to check beyond that

**Review by the-mikedavis — APPROVED** [review]

**Review by pascalkuthe — APPROVED** [review]

**Pull request merged**